x-pack/plugins/proxy/server/proxy.ts

```diff
@@ -0,0 +1,241 @@
+        readFileAsync(config.ca),
+      ]);
+    } catch (err) {
+      this.log.fatal('Unable to read SSL cerificate information', err);
+      throw new Error('You must provide valid paths for cert, key and ca');
+    }
+
+    this.httpsAgent = new HTTPSAgent({
+      keepAlive: true,
+      cert: tlsCert,
+      key: tlsKey,
+      ca: tlsCa,
+    });
+
+    const ssl = {
+      enabled: true,
+      redirectHttpFromPort: this.port,
+      certificate: tlsCert.toString(),
+      key: tlsKey.toString(),
+      certificateAuthorities: [tlsCa.toString()],
+      cipherSuites: config.cipherSuites,
+      keyPassphrase: undefined,
+      supportedProtocols: config.supportedProtocols,
+      requestCert: true,
+    };
+    return ssl;
+  }
+
+  private setConfig(config: ProxyPluginType) {
+    this.port = config.port;
+    this.maxRetry = config.maxRetry;
+    this.requestBackoff = config.requestBackoff;
+  }
+
+  public async start() {
+    await this.clusterDocClient.start();
+    const start: ProxyServiceStart = {
+      assignResource: this.assignResource.bind(this),
+      unassignResource: this.unassignResource.bind(this),
+      proxyResource: this.proxyResource.bind(this),
+      proxyRequest: this.proxyRequest.bind(this),
+      getAllocation: this.getAllocation.bind(this),
+    };
+    return start;
+  }
+
+  public async stop() {
+    await this.clusterDocClient.stop();
+  }
+
+  public async assignResource(
+    resource: string,
+    type: string,
+    state: RouteState,
+    node?: string
+  ): Promise<void> {
+    await this.clusterDocClient.assignResource(resource, type, state, node);
+  }
+
+  public async unassignResource(resource: string) {
+    await this.clusterDocClient.unassignResource(resource);
+  }
+
+  public proxyResource(resource: string): (req: KibanaRequest) => Promise<any> {
+    return (req: KibanaRequest) => {
+      return this.proxyRequest(req, resource);
+    };
+  }
+
+  // @TODO update to allow passing of request parametsrs
+  public async proxyRequest(req: KibanaRequest, resource?: string, retryCount = 0): Promise<any> {
+    const method = req.route.method;
+    const url = new URL(req.url.toString());
+    const headers = req.headers;
+    const body = req.body;
+    resource = resource || url.pathname;
+    const node = await this.clusterDocClient.getNodeForResource(resource);
+
+    if (!node || node.state === RouteState.Closed) {
+      const msg = `No node was found for resource ${resource}`;
+      this.log.debug(msg);
+      throw new Error(msg);
+    }
+
+    if (node.state === RouteState.Initializing) {
+      throw new Error(`Node ${node.node} is initializing`);
+    }
+    if (node.state === RouteState.Closing) {
+      throw new Error(`Node ${node.node} is closing.`);
+    }
+
+    url.hostname = node.node;
+    try {
+      const opts = {
+        headers,
+        payload: body,
+      };
+      const res = await this.wreck.request(method, url.toString(), opts);
```

Comment:
i had thought about that but worried it would complicate future use of this api (other consumers might manipulate the reply before sending it back?)

If we feel strongly here i'm happy to move to streams.